// antd-editable/es/Cell.js
import "antd/es/input/style/css";
import _Input from "antd/es/input";
import "antd/es/select/style/css";
import _Select from "antd/es/select";
import "antd/es/form/style/css";
import _Form from "antd/es/form";

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

import React, { useEffect } from 'react';
var Item = _Form.Item;
var Option = _Select.Option;

var Cell = function Cell(_ref) {
  var form = _ref.form,
      dataIndex = _ref.dataIndex,
      rowIndex = _ref.rowIndex,
      curCell = _ref.curCell,
      onSetCurCell = _ref.onSetCurCell,
      initialValue = _ref.initialValue,
      _ref$rules = _ref.rules,
      rules = _ref$rules === void 0 ? [] : _ref$rules,
      isSelect = _ref.isSelect;
  // 是否处于可编辑状态
  var isEditing = !!curCell && curCell.dataIndex === dataIndex && curCell.rowIndex === rowIndex; // input的Ref，用于激活后的focus

  var inputRef = useFocus(isEditing); // 将当前的Cell激活

  function handleSetCurCell() {
    if (curCell) {
      form.validateFields(["".concat(curCell.dataIndex, "-").concat(curCell.rowIndex)], function (err) {
        if (!err) onSetCurCell({
          dataIndex: dataIndex,
          rowIndex: rowIndex
        });
      });
    } else {
      onSetCurCell({
        dataIndex: dataIndex,
        rowIndex: rowIndex
      });
    }
  } // 保存值到表单域里


  function handleSave() {
    form.validateFields(["".concat(dataIndex, "-").concat(rowIndex)], function (err) {
      if (!err) {
        onSetCurCell(null);
      }
    });
  } // 静态值


  var stockCell = React.createElement("div", {
    onClick: handleSetCurCell,
    className: "editable-cell-value-wrap"
  }, Array.isArray(isSelect) ? function foo() {
    var target = isSelect.find(function (_ref2) {
      var value = _ref2.value;
      return value === initialValue;
    });
    return target ? target.label : initialValue;
  }() : initialValue); // 获取控件

  var getFormItem = function getFormItem() {
    if (isSelect) {
      return React.createElement(_Select, {
        onChange: handleSave,
        ref: inputRef,
        onBlur: handleSave,
        style: {
          width: '100%'
        }
      }, Array.isArray(isSelect) && isSelect.map(function (item, index) {
        return React.createElement(Option, {
          key: index,
          value: item.value
        }, item.label);
      }));
    } else {
      return React.createElement(_Input, {
        ref: inputRef,
        onPressEnter: handleSave,
        onBlur: handleSave
      });
    }
  };

  var rulesWithCellInfo = rules.map(function (item) {
    var _validator = item.validator;
    return _validator ? _extends({}, item, {
      validator: function validator(rule, value, callback) {
        _validator(rule, value, callback, curCell);
      }
    }) : item;
  });
  return React.createElement("div", {
    style: {
      textAlign: 'left'
    }
  }, React.createElement(Item, null, form.getFieldDecorator("".concat(dataIndex, "-").concat(rowIndex), {
    initialValue: initialValue === '--' ? '' : initialValue,
    rules: rulesWithCellInfo
  })(isEditing ? getFormItem() : stockCell)));
};

var useFocus = function useFocus(isEditing) {
  var inputRef = React.createRef();
  useEffect(function () {
    if (isEditing) {
      if (inputRef.current) inputRef.current.focus();
    }
  }, [isEditing]);
  return inputRef;
};

export default Cell;
