// antd-editable/lib/Editable.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

require("antd/es/form/style/css");

var _form = _interopRequireDefault(require("antd/es/form"));

require("antd/es/button/style/css");

var _button = _interopRequireDefault(require("antd/es/button"));

require("antd/es/table/style/css");

var _table = _interopRequireDefault(require("antd/es/table"));

var _react = _interopRequireDefault(require("react"));

var _useProps2 = _interopRequireDefault(require("./useProps"));

var _EditableWrapper = _interopRequireDefault(require("./style/EditableWrapper"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

var __rest = void 0 && (void 0).__rest || function (s, e) {
  var t = {};

  for (var p in s) {
    if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0) t[p] = s[p];
  }

  if (s != null && typeof Object.getOwnPropertySymbols === "function") for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
    if (e.indexOf(p[i]) < 0) t[p[i]] = s[p[i]];
  }
  return t;
};

var Editable = function Editable(_a) {
  var _a$dataSource = _a.dataSource,
      dataSource = _a$dataSource === void 0 ? [] : _a$dataSource,
      _a$columns = _a.columns,
      columns = _a$columns === void 0 ? [] : _a$columns,
      _a$onCellChange = _a.onCellChange,
      onCellChange = _a$onCellChange === void 0 ? function () {} : _a$onCellChange,
      _a$btnProps = _a.btnProps,
      btnProps = _a$btnProps === void 0 ? {
    style: {
      marginTop: 10
    }
  } : _a$btnProps,
      _a$btnText = _a.btnText,
      btnText = _a$btnText === void 0 ? '保存' : _a$btnText,
      form = _a.form,
      pagination = _a.pagination,
      onSubmit = _a.onSubmit,
      resProps = __rest(_a, ["dataSource", "columns", "onCellChange", "btnProps", "btnText", "form", "pagination", "onSubmit"]);

  var _useProps = (0, _useProps2["default"])(dataSource, columns, onCellChange, form),
      cacheSource = _useProps.cacheSource,
      editColumns = _useProps.editColumns,
      hasError = _useProps.hasError;

  function handleSubmit() {
    if (onSubmit) onSubmit(cacheSource);
  }

  return _react["default"].createElement(_EditableWrapper["default"], null, _react["default"].createElement(_table["default"], _extends({
    className: "metro-editable",
    dataSource: cacheSource,
    columns: editColumns,
    rowClassName: function rowClassName() {
      return 'metro-editable-row';
    },
    pagination: pagination || false
  }, resProps)), onSubmit && _react["default"].createElement(_button["default"], _extends({
    onClick: handleSubmit,
    disabled: hasError,
    className: "metro-editable-submitBtn"
  }, btnProps), btnText));
};

var _default = _form["default"].create()(Editable);

exports["default"] = _default;
